aten/src/ATen/native/cuda/TensorShape.cu

```diff
@@ -268,6 +282,305 @@ std::pair<at::Tensor, std::vector<int64_t*>> pack_vecs(
+// `src` size (i.e., actual_chunk_size) is less than `max_chunk_size`.
+static __device__ __inline__ void copy_chunk_with_pad(
+  char* dst,
+  const char* src,
+  int64_t max_chunk_size,
+  int64_t actual_chunk_size,
+  int64_t thread_idx,
+  int64_t num_threads
+) {
+  // Fast path when the number of threads is larger than the number of bytes to be copied
+  // (i.e., max_chunk_size). In this case, each thread only copies 1 byte.
+  // For 0 <= thread_idx < actual_chunk_size, the thread copies data from `src`.
+  // For actual_chunk_size <= thread_idx < max_chunk_size, the thread set the val=0 for padding.
+  if (max_chunk_size < num_threads) {
+    int64_t val = 0;
+    if (thread_idx < actual_chunk_size) {
+      val = src[thread_idx];
+    }
+    if(thread_idx < max_chunk_size) {
+      dst[thread_idx] = val;
+    }
+    return;
+  }
+  uint4 zero = get_zero_uint4();
+  // Split dst array into three parts:
+  // [dst, dst+align_off), [dst+align_off, dst+align_end), [dst+align_end, dst+max_chunk_size)
+  // The second part is aligned with BYTES_PER_THREAD(=16 bytes) to enable `stream_store128`.
+  int64_t align_off, aligned_size;
+  get_aligned_region(dst, actual_chunk_size, BYTES_PER_THREAD, align_off, aligned_size);
+  int64_t align_end = align_off + aligned_size;
+  for (
+    int64_t i = align_off + thread_idx * BYTES_PER_THREAD;
+    i < align_end;
+    i += num_threads * BYTES_PER_THREAD
+  ) {
+    uint4 val = zero;
+    if(is_aligned<uint4>(src + i)) {
+      stream_load128(val, src + i);
+    } else {
+      for (size_t j = 0; j < BYTES_PER_THREAD; ++j) {
+        reinterpret_cast<char*>(&val)[j] = src[i + j];
+      }
+    }
+    stream_store128(&dst[i], val);
+  }
+  // Copy data for the first part of dst array [dst, dst+align_off).
+  // Check `thread_idx<max_chunk_sze` for the edge case that max_chunk_size < align_off.
+  if(thread_idx < align_off && thread_idx < max_chunk_size) {
+    char val = (char) 0;
+    if (thread_idx < actual_chunk_size) {
+      val = src[thread_idx];
+    }
+    dst[thread_idx] = val;
+  }
+  // Copy data for the third part of dst array [dst+align_end, dst+max_chunk_size).
+  while(align_end + thread_idx < max_chunk_size) {
+    char val = (char) 0;
+    if (align_end + thread_idx < actual_chunk_size) {
+      val = src[align_end + thread_idx];
+    }
+    dst[align_end + thread_idx] = val;
+    align_end += num_threads;
+  }
+}
+
+// Read `actual_chunk_size` bytes from bfloat16 array `src`  (i.e., actual_chunk_size/2 bfloat16 elements),
+// cast bfloat16 to float, and write to float array `dst`. Pad 0.0f if 2*actual_chunk_size < max_chunk_size.
+// src: pointer to a bfloat16 array
+// dst: pointer to a float32 array
+// max_chunk_size: number of bytes in dst array
+// actual_chunk_size: number of bytes in src array, which is less than or equal to max_chunk_size / 2
+static __device__ __inline__ void copy_chunk_with_pad_cast_bfloat16_to_float(
```

Comment:
This path is a bit specific. Can we make it a template function? I think it's doable with https://github.com/pytorch/pytorch/blob/main/c10/util/TypeCast.h

Reply:
>  To avoid some numerical issues, I will use intrinsics here when available.

Hmm I don't think that's necessary. IIUC calling `static_cast<float>` on `at::BFloat16` should already be dispatching to an intrinic: https://github.com/pytorch/pytorch/blob/main/c10/util/BFloat16-inl.h#L41